ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,339 @@
+     */
+    FREE,
+    /**
+     * The LogSegment is currently being written to. It still has capacity remaining for more writes.
+     */
+    ACTIVE,
+    /**
+     * The LogSegment is full and is accepting no more writes.
+     */
+    SEALED
+  }
+
+  private final FileChannel fileChannel;
+  private final File file;
+  private final long capacityInBytes;
+  private final String name;
+  private final Pair<File, FileChannel> segmentView;
+  private final StoreMetrics metrics;
+  private final AtomicLong endOffset;
+  private final AtomicLong refCount = new AtomicLong(0);
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  /**
+   * The state of the segment.
+   * <p/>
+   * The natural state transition of a LogSegment is FREE -> ACTIVE -> SEALED but it can be externally manipulated to
+   * move in any direction.
+   */
+  State state;
+  /**
+   * Reference to the segment that logically "follows" this segment.
+   */
+  LogSegment next = null;
+
+  /**
+   * Creates a LogSegment abstraction.
+   * @param name the desired name of the segment.
+   * @param file the backing file for this segment.
+   * @param capacityInBytes the intended capacity of the segment
+   * @param metrics the {@link StoreMetrics} instance to use.
+   * @throws IOException if the file cannot be read or created
+   */
+  LogSegment(String name, File file, long capacityInBytes, StoreMetrics metrics)
+      throws IOException {
+    if (!file.exists() || !file.isFile()) {
+      throw new IllegalArgumentException(file.getAbsolutePath() + " does not exist or is not a file");
+    }
+    this.file = file;
+    this.name = name;
+    this.capacityInBytes = capacityInBytes;
+    this.metrics = metrics;
+    fileChannel = Utils.openChannel(file, true);
+    segmentView = new Pair<>(file, fileChannel);
+    // state is always initialized to FREE and the end offset is always initialized to 0.
+    // externals will set the correct values for these two variables
+    state = State.FREE;
+    endOffset = new AtomicLong(0);
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Attempts to write the {@code buffer} in its entirety in this segment. To guarantee that the write is persisted,
+   * {@link #flush()} has to be called.
+   * <p/>
+   * The write is not started if it cannot be completed.
+   * @param buffer The buffer from which data needs to be written from
+   * @return the number of bytes written.
+   * @throws IllegalArgumentException if there is not enough space for {@code buffer}
+   * @throws IllegalStateException if the segment is not {@link State#ACTIVE}.
+   * @throws IOException if data could not be written to the file because of I/O errors
+   */
+  @Override
+  public int appendFrom(ByteBuffer buffer)
+      throws IOException {
+    int bytesWritten = 0;
+    ensureActive();
+    if (endOffset.get() + buffer.remaining() > capacityInBytes) {
+      metrics.overflowWriteError.inc();
+      throw new IllegalArgumentException(
```

Comment:
I don't think this is easy to call - my thinking is that this is an illegal argument because the segment exposes how much capacity it has left and if someone sent in a buffer of a higher size, it is probably an illegal argument? Anyway, not inclined either way - if you feel strongly about it, can change.